src/envoy/http/service_control/client_cache_test.cc

```diff
@@ -268,6 +283,184 @@ TEST_F(ClientCacheQuotaResponseErrorTypeTest, ApiKeyInvalid) {
+                                          Envoy::Tracing::Span&,
+                                          HttpCall::DoneFunc on_done) {
+        http_done = on_done;
+        return http_call_.get();
+      }));
+  EXPECT_CALL(*http_call_, call()).Times(1);
+  injectFactoryMocks();
+
+  // Run function under test.
+  const CheckRequest request = getValidCheckRequest();
+  cache_->callCheck(
+      request, mock_parent_span_,
+      absl::bind_front(&ClientCacheCheckHttpRequestTest::verifyCheckDone, this,
+                       Code::OK));
+
+  // RPC is pending, no callback invoked until http is done.
+  EXPECT_EQ(got_num_callbacks, 0);
+
+  // Stimulate successful http response.
+  // Test tear down will check the check callback is invoked.
+  want_num_callbacks++;
+  std::string response_body;
+  const CheckResponse response = getValidCheckResponse();
+  response.SerializeToString(&response_body);
+  http_done(Status::OK, response_body);
+
+  // Check stats.
+  checkAndReset(stats_base_.stats().check_.OK_, 1);
+}
+
+TEST_F(ClientCacheCheckHttpRequestTest, OneHttpCallWithBadBody) {
+  // Only 1 http call will be created.
+  HttpCall::DoneFunc http_done;
+  EXPECT_CALL(*check_call_factory_, createHttpCall(_, _, _))
+      .WillOnce(Invoke([this, &http_done](const Envoy::Protobuf::Message&,
+                                          Envoy::Tracing::Span&,
+                                          HttpCall::DoneFunc on_done) {
+        http_done = on_done;
+        return http_call_.get();
+      }));
+  EXPECT_CALL(*http_call_, call()).Times(1);
+  injectFactoryMocks();
+
+  // Run function under test.
+  const CheckRequest request = getValidCheckRequest();
+  cache_->callCheck(
+      request, mock_parent_span_,
+      absl::bind_front(&ClientCacheCheckHttpRequestTest::verifyCheckDone, this,
+                       Code::INTERNAL));
+
+  // RPC is pending, no callback invoked until http is done.
+  EXPECT_EQ(got_num_callbacks, 0);
+
+  // Stimulate bad http response body.
+  // Test tear down will check the check callback is invoked.
+  want_num_callbacks++;
+  http_done(Status::OK, "this http body does not parse into a CheckResponse");
+
+  // Check stats.
+  checkAndReset(stats_base_.stats().check_.INVALID_ARGUMENT_, 1);
+  // TODO(nareddyt): This should probably be treated as a control plane fault.
+  checkAndReset(stats_base_.stats().filter_.denied_producer_error_, 1);
+}
+
+TEST_F(ClientCacheCheckHttpRequestTest, OnePendingHttpCallCancelled) {
+  // Only 1 http call will be created.
+  HttpCall::DoneFunc http_done;
+  EXPECT_CALL(*check_call_factory_, createHttpCall(_, _, _))
+      .WillOnce(Invoke([this, &http_done](const Envoy::Protobuf::Message&,
+                                          Envoy::Tracing::Span&,
+                                          HttpCall::DoneFunc on_done) {
+        http_done = on_done;
+        return http_call_.get();
+      }));
+  EXPECT_CALL(*http_call_, call()).Times(1);
+  injectFactoryMocks();
```

Comment:
Looks like line 428 to 437 are the same for all test case,  maybe move into the test class Setup

Reply:
Done